### PR TITLE
chore: clean up product mixin

### PR DESCRIPTION
**Pull request by tobi-or-not-tobi**
- fixed issue with default value for product
- drop the usage of value state in the product package

### CODE REVIEW

**Review by dunqan — DISMISSED** [review]

File: libs/domain/cart/add/src/add.component.ts [hunk 1]
```diff
@@ -21,7 +29,9 @@ export class CartAddComponent extends ProductMixin(
   static styles = styles;
 
   protected cartService = resolve(CartService);
+
   @state() protected isInvalid = false;
+  protected isInvalid2 = signal(false);
```
Comment (dunqan) [inline libs/domain/cart/add/src/add.component.ts]: 👀

File: libs/domain/cart/add/src/add.component.ts [hunk 2]
```diff
@@ -58,22 +68,29 @@ export class CartAddComponent extends ProductMixin(
...
-    if (this.$product()) {
-      this.shadowRoot
-        ?.querySelector<QuantityInputComponent>('oryx-cart-quantity-input')
-        ?.reset?.();
-    }
-    return value;
+  @elementEffect()
+  protected reset = effect(() => {
+    if (this.$product()) this.input?.reset?.();
+  });
```
Comment (dunqan) [inline libs/domain/cart/add/src/add.component.ts]:
```suggestion
  protected reset = () => {
    if (this.$product()) this.input?.reset?.();
  };
```

File: libs/domain/product/src/mixins/product.mixin.ts [hunk 3]
```diff
@@ -34,15 +30,23 @@ export const ProductMixin = <
...
+    protected productService = resolve(ProductService, null);
+    protected contextController = new ContextController(this);
+
+    protected $context = signal(
+      this.contextController.get<string>(ProductContext.SKU)
+    );
 
-    @asyncState()
-    protected product = valueType(this.productController.getProduct());
+    protected setProductContext = effect(() => {
```
Comment (dunqan) [inline libs/domain/product/src/mixins/product.mixin.ts]: `@elementEffect`?

File: libs/domain/product/src/mixins/product.mixin.ts [hunk 4]
```diff
@@ -34,15 +30,23 @@ export const ProductMixin = <
   class ProductMixinClass extends superClass {
     @signalProperty({ reflect: true }) sku?: string;
 
-    protected productController = new ProductController(this);
+    protected productService = resolve(ProductService, null);
+    protected contextController = new ContextController(this);
+
+    protected $context = signal(
```
Comment (dunqan) [inline libs/domain/product/src/mixins/product.mixin.ts]: `$context` could be confusing here, as it's too generic term... what about `$sku`? (I'm aware we already have sku property)...

File: libs/domain/product/src/mixins/product.mixin.ts [hunk 5]
```diff
@@ -34,15 +30,23 @@ export const ProductMixin = <
...
+      this.contextController.get<string>(ProductContext.SKU)
+    );
 
-    @asyncState()
-    protected product = valueType(this.productController.getProduct());
+    protected setProductContext = effect(() => {
+      if (this.sku && !this.$context()) {
```
Comment (dunqan) [inline libs/domain/product/src/mixins/product.mixin.ts]: I'm wondering if we should check for `!this.$context()` as my reasoning is, that if we pass `sku` as a property, we should override context (from parent element)

**Review by dunqan — CHANGES_REQUESTED** [review]

File: libs/domain/cart/add/src/add.component.ts (hunk 2 above)
Comment (tobi-or-not-tobi) [reply libs/domain/cart/add/src/add.component.ts]: We need to reset the input whenever the SKU is changing

File: libs/domain/product/src/mixins/product.mixin.ts [hunk 6]
```diff
@@ -34,15 +30,25 @@ export const ProductMixin = <
...
+      const sku = this.$productContext() || this.sku;
+      return sku ? this.productService?.get({ sku }) : of(null);
     });
+
+    @elementEffect()
+    protected setProductContext = () => {
+      if (this.sku && this.$productContext() !== this.sku) {
+        this.contextController.provide(ProductContext.SKU, this.sku);
+      }
+    };
```
Comment (supproduction) [inline libs/domain/product/src/mixins/product.mixin.ts]: is it correct that we should provide context sku inside mixin?
In this case we are going to pollute dom

As far I know we need it only for card component, don't we?

**Review by supproduction — DISMISSED** [review]

File: libs/template/labs/src/bazaarvoice/rating/rating.component.ts [hunk 7]
```diff
@@ -12,22 +11,19 @@ export class BazaarvoiceRatingComponent extends ProductMixin(
...
-      if (product) {
-        loadBvScript().then((bv) => {
-          const summaryRefId = `${this.uid}-average-rating`;
-          this.createContainer(summaryRefId);
-
-          bv.ui('rr', 'show_reviews', {
-            productId: product?.sku,
-            summaryContainerDiv: summaryRefId,
-          });
+  protected loadScript = effect(() => {
```
Comment (dunqan) [inline libs/template/labs/src/bazaarvoice/rating/rating.component.ts]:
```suggestion
  @elementEffect()
  protected loadScript = () => {
```

File: libs/template/labs/src/bazaarvoice/review-list/review-list.component.ts [hunk 8]
```diff
@@ -13,23 +12,19 @@ export class BazaarvoiceReviewListComponent extends ProductMixin(
...
-        loadBvScript().then((bv) => {
-          this.hideContainer();
-          const reviewsRefId = `${this.uid}-reviews`;
-          this.createContainer(reviewsRefId);
-
-          bv.ui('rr', 'show_reviews', {
-            productId: product?.sku,
-            contentContainerDiv: reviewsRefId,
-          });
+  protected loadScript = effect(() => {
```
Comment (dunqan) [inline libs/template/labs/src/bazaarvoice/review-list/review-list.component.ts]:
```suggestion
  @elementEffect()
  protected loadScript = () => {
```

**Review by dunqan — APPROVED** [review]

**Pull request merged**